README.md
# CMake 使用及常见问题

# 环境检查 
检查是否安装**cmke**
检查是否安装**python3**

# 使用方法
### 使用build.sh脚本
- 1 根目录下执行`./build.sh build`,这个命令会建立build文件夹并执行**cmake**命令，生成makefile文件，如果执行失败请检查**build.sh**是否有可执行权限,添加可执行权限。
    ```shell
    chmod +x build.sh && ./build.sh build
    ```
- 2 `./build.sh build`执行成功后执行命令，这个命令会使用上一步生成的makefile文件生成最终的可执行文件**flashimage.bin**。
    ```shell
    ./build.sh all
    ```
  
- 3 执行`./build.sh clean`命令会清除**build**文件夹
    ```shell
    ./build.sh clean
    ```

### 直接使用cmake

- 1 在项目根目录下建立并进入**build**文件夹
    ```shell
    mkdir build && cd build
    ```
- 2 执行**cmake**命令
    ```shell
    cmake -DPROJECT=wifi_mcu_develop ..
    ```
- 3 生成**Makefile**文件之后执行`make`命令生成可执行文件，可执行文件**flashimage.bin**在**build/bin**文件夹下
    ```shell
    make
    ```


# 常见问题

### 1 执行`./build.sh build`后出现CMAKE_CXX_COMPILER not set

检查是否安装**g++**


### 2 tools/Linux/arm-none-eabi-gcc/bin/arm-none-eabi-gcc: Permission denied
出现这个问题很大原因是这个可执行文件没有可执行权限，并且除`tools/Linux/arm-none-eabi-gcc/bin`之外其他文件也可能没有可执行权限，因此可对`tools/Linux/arm-none-eabi-gcc/`文件夹以及相关子目录添加可执行权限。




CHANGELOG.md

<table>
    <tr>
        <td>日期</td>
        <td>当前SDK版本</td>
        <td>上游版本</td>
        <td>摘要</td>
    </tr>
    <tr>
        <td>
            2022-11-25 15:00:00
        </td>
        <td>
            v1.7
        </td>
        <td>
            73bf2b87
        </td>
        <td>
            1. 修复链接脚本问题（配置默认小端格式，修复 app 分区可能越界的问题） <br />
            2. 增加亚马逊 aws 连接组件和 demo 工程 <br />
            3. 修复上次解决复位后 PHY 异常问题引入的单 ble 模式不工作的问题，在 PHY 的初始化前后，开关 RF_TRX <br />
            4. 修复低功耗起来后频点偏差的问题 <br />
            5. AT 组件增加 dump 参数的调试接口 <br />
            6. 增加 AP 模式下扫描的功能 <br />
            7. 修复硬件过滤器多路配置的同步问题 <br />
            8. WiFi 组件增加 MAC_STA_STARTUP、MAC_AP_STARTUP 状态 <br />
            9. WiFi TX 功率补偿接口支持 b 模式 gn 模式单独补偿 <br />
            10. 改善低温（-40~-10）下 EVM 指标 <br />
            11. 增加腾讯云 qcloud 组件和 demo <br />
            12. 修复 WiFi disconnect 的时候会锁定信道的问题 <br />
            13. 修复 WiFi PA 误开关的问题 <br />
            14. 更新低功耗策略，WiFi disconnect 之后默认不允许 CPU 休眠 <br />
            15. 移除 cJSON 组件，版本太旧，且跟新版本 API 不兼容，顾移除。如果用户业务或者组件需要使用 cJSON，请添加到该业务代码或者组件里自行维护 <br />
            16. 修复速率集相关代码重构引入的 STA 模式连接 11b only 路由器会失败的问题 <br />
            17. 修复 STA 关联上路由器后 scan 操作可能的死机问题 <br />
        </td>
    </tr>
    <tr>
        <td>
            2022-10-12 10:00:00
        </td>
        <td>
            v1.6
        </td>
        <td>
            c5088226
        </td>
        <td>
            1. WIFI Lib 版本更新为 v1.0.9 <br />
            2. SDK 版本更新为 v1.6 <br />
             <br /><b>SDK changelog:</b> <br />
            1. 优化light_sleep模式下功耗，修复该模式下OSTick补偿方式(24小时误差小于15S） <br />
            2. 修复PWM Demo示例工程中不能配置引脚功能的问题 <br />
            3. 增强AT命令(wifi scan)的健壮性 <br />
            4. 更新<b>wifilib</b> <br />
				4.1. 优化内存，增加用户可用内存(os_heap)33KB+ <br />
				4.2. PLL校准只在温度变化跨度较大情况下进行，降低高温区由于温区边界宽度较小反复触发PLL校准的问题. <br />
				4.3. 修复 WiFi scan 在扫描指定信道的时候可能使用内存脏数据的问题 <br />
				4.4. 增加指定信道扫描功能（当 STA 处于连接状态，不允许指定信道扫描） <br />
				4.5. RFDC校准增强健壮性 <br />
				4.6. 优化配置，修复在某特定场景下的CCA判断错误的问题，增强抗干扰特性 <br />
				4.7. SoftAP启动时增加随机ms延时 <br />
				4.8. 优化PHY的配置时间，减小light_sleep模式下tbtt提前唤醒时间 <br />
				4.9. 开启monitor模式时可以自动信道切换进行scan操作，保持smartcfg配网(usr_sniffer)模式下信道切换交给用户操作 <br />
        </td>
    </tr>
    <tr>
        <td>
            2022-09-07 10:00:00
        </td>
        <td>
            v1.5
        </td>
        <td>
            1d029698
        </td>
        <td>
            1. WIFI 版本更新为 v1.0.8 <br />
            2. SDK 版本更新为 v1.5 <br />
             <br /><b>SDK changelog:</b> <br />
            1. 增加 GPIO 测试 Demo <br />
            2. 修改 ramcode_dl 的复位接口<br />
            3. 修改 pll 周期校准启动时机     <br />
            4. 添加读取 Flash UID 接口 <br />
            5. 修复 ap 模式下发送 send_null_frame_to_AP 函数的问题 <br />
            6. 优化 lwip 配置 <br />
            7. 增加设置 erp 类型和 rts threshold 的接口<br />
            8. 修复 dhcpd 任务被多次创建同时运行的问题 <br />
            9. AP 模式默认使用高功率表  <br />
            10. 重新设计速率集配置，将 bg 速率集移动到 wifi_port.c/h 以供用户自行配置<br />
            11. 修复 sniffer mem pool 被反复初始化的问题 <br />
            13. 增加软件 duration 配置失能接口，默认使能软件配置功能，调用该接口后失能
        </td>
    </tr>
<tr>
        <td>
            2022-07-15 10:50:00
        </td>
        <td>
            v1.4
        </td>
        <td>
            288ee4c6
        </td>
        <td>
            1. WIFI 版本更新为 v1.0.7 <br />
            2. BLE 版本更新为 v1.0.4 <br />
            3. SDK 版本更新为 v1.4 <br />
             <br /><b>SDK changelog:</b> <br />
            1. 新增 RAMCODE 命令，增加获取 flash 信息和 Flash 整片测试命令 <br />
            2. at 指令增加 nvic reset 功能<br />
            3. 优化 wifi 网络性能<br />
            4. 修复 connect 的时候 scan 到目标网络 但是没有发起 join 的问题     <br />
            5. img 校准使用的 RAM 区域 4 字节对齐 <br />
            6. wifi rf 校准优化 <br />
            7. 增加 BLE 硬件复位接口 <br />
            8. 增加 BLE 边带测试命令 <br />
            9. 修复在重关联的时候拿到 ip 后偶现 dhcp release 的问题 <br />
            10. wifi 校准函数合并为一个 wifi_rf_calibration
    原来为 wifi_rf_preprocess 和 wifi_rf_image_cal 
            <br />
        </td>
    </tr>
<tr>
        <td>
            2022-06-24 10:50:00
        </td>
        <td>
            no version
        </td>
        <td>
            33ed5a55
        </td>
        <td>
            1. 优化 mbedtls 内存. MBEDTLS_SSL_OUT_CONTENT_LEN 设置为4K <br />
            2. 优化 wifi 传输速率 <br />
            3. BLE 示例工程 sct 更新 <br />
            4. add battery release project <br />
            5. add mouse release project <br />
            6. 添加接口 wifi_softap_set_max_supp_sta_num,wifi_softap_get_max_supp_sta_num 可以设置/获取最大支持的sta数目 <br />
        </td>
    </tr>
    <tr>
        <td>
            2022-06-01 20:55:00
        </td>
        <td>
            no version
        </td>
        <td>
            c2382740
        </td>
        <td>
            1. 增加 single tone 调试 AT+PVTCMD=s_tone_mode BLE AT+PVTCMD=s_tone_mode WIFI <br />
            2. 修复在未设置 got ip function callback 的时候 WiFi 无法进入睡眠的问题 <br />
            3. 增加 AT OTA 命令程序 URL 缓存大小；增加 AT CMD 缓存大小； <br />
            4. 增加 https client path 和 CONTENT 缓冲区大小； <br />
            5. mbedtls 移除对 MBEDTLS_SSL_IN_CONTENT_LEN 和 MBEDTLS_SSL_OUT_CONTENT_LEN 的配置，继承 MBEDTLS_SSL_MAX_CONTENT_LEN。 <br />
        </td>
    </tr>
    <tr>
        <td>
            2022-05-25 16:45:00
        </td>
        <td>
            no version
        </td>
        <td>
            75faa75d
        </td>
        <td>
            1. 修复在未设置 got ip function callback 的时候 WiFi 无法进入睡眠的问题 <br />
            2. 射频测试的功能更新，增加 s_tone_mode 命令 <br />
        </td>
    </tr>
    <tr>
        <td>
            2022-05-24 18:04:00
        </td>
        <td>
            no version
        </td>
        <td>
            75faa75d
        </td>
        <td>
            1. 修复 log 输出参数不匹配问题及其他编译警告 <br />
            2. 不定参使用 ##__VA_ARGS__ 来避免无参数错误 <br />
            3. GNU 下，使用 __attribute__((format(printf, 2, 3))) 和 -Wformat=2 增强对 printf 类型的不定参函数进行 format 和 参数匹配检查 <br />
            4. 更新 boot 程序，并添加 AT 指令 AT+BLE_START 以允许跳转到 OTA 分区执行代码 <br />

mcu/ln882h/ln882h.h
#ifndef __LN882H_H__
#define __LN882H_H__

#ifdef __cplusplus
extern "C" {
#endif


/* -------------------------  Interrupt Number Definition  ------------------------ */
typedef enum IRQn
{
    /* -------------------  Processor Exceptions Numbers  ----------------------------- */
    Reset_IRQn              = -15,              /* -15  Reset Vector, invoked on Power up and warm reset                       */
    NMI_IRQn                = -14,              /* -14  Non maskable Interrupt, cannot be stopped or preempted                           */
    HardFault_IRQn          = -13,              /* -13  Hard Fault, all classes of Fault                                       */
    MemoryManagement_IRQn   = -12,              /* -12  Memory Management, MPU mismatch, including Access Violation and No Match */
    BusFault_IRQn           = -11,              /* -11  Bus Fault, Pre-Fetch-, Memory Access Fault, other address/memory
                                                        related Fault                                                          */
    UsageFault_IRQn         = -10,              /* -10  Usage Fault, i.e. Undef Instruction, Illegal State Transition          */
    SVCall_IRQn             =  -5,              /* -5   System Service Call via SVC instruction                                */
    DebugMonitor_IRQn       =  -4,              /* -4   Debug Monitor                                                          */
    PendSV_IRQn             =  -2,              /* -2   Pendable request for system service                                    */
    SysTick_IRQn            =  -1,              /* -1   System Tick Timer                                                      */

    /* -------------------  Processor Interrupt Numbers  ------------------------------ */
    WDT_IRQn                = 0,
    EXT_IRQn                = 1,
    RTC_IRQn                = 2,
    RFSLP_IRQn              = 3,
    MAC_IRQn                = 4,
    BLE_WAKE_IRQn           = 5,
    BLE_ERR_IRQn            = 6,
    BLE_MAC_IRQn            = 7,    
    DMA_IRQn                = 8,
    QSPI_IRQn               = 9,
    SDIO_1_IRQn             = 10,
    SDIO_2_IRQn             = 11,
    SDIO_3_IRQn             = 12,   
    FPIXC_IRQn              = 13,
    FPOFC_IRQn              = 14,
    FPUFC_IRQn              = 15,
    FPIOC_IRQn              = 16,
    FPDZC_IRQn              = 17,
    FPIDC_IRQn              = 18,
    I2C_IRQn                = 19,
    SPI0_IRQn               = 20,
    SPI1_IRQn               = 21,  
    UART0_IRQn              = 22,
    UART1_IRQn              = 23,
    UART2_IRQn              = 24, 
    ADC_IRQn                = 25,
    WS2811_IRQn             = 26, 
    I2S_IRQn                = 27,
    GPIOA_IRQn              = 28,
    GPIOB_IRQn              = 29,
    TIMER0_IRQn             = 30,
    TIMER1_IRQn             = 31,
    TIMER2_IRQn             = 32,
    TIMER3_IRQn             = 33,
    ADV_TIMER_IRQn          = 34,
    AES_IRQn                = 35,
    TRNG_IRQn               = 36,
    PAOTD_IRQn              = 37,  

} IRQn_Type;

/* ================================================================================ */
/* ================      Processor and Core Peripheral Section     ================ */
/* ================================================================================ */

#define __CM4_REV                 0x0201U    /*!< Core Revision r2p1 */
/* ToDo: define the correct core features for the <Device> */
#define __MPU_PRESENT             1U         /*!< Set to 1 if MPU is present */
#define __VTOR_PRESENT            1U         /*!< Set to 1 if VTOR is present */
#define __NVIC_PRIO_BITS          3U         /*!< Number of Bits used for Priority Levels */
#define __Vendor_SysTickConfig    0U         /*!< Set to 1 if different SysTick Config is used */
#define __FPU_PRESENT             1U         /*!< Set to 1 if FPU is present */
#define __FPU_DP                  0U         /*!< Set to 1 if FPU is double precision FPU (default is single precision FPU) */
#define __ICACHE_PRESENT          0U         /*!< Set to 1 if I-Cache is present */
#define __DCACHE_PRESENT          0U         /*!< Set to 1 if D-Cache is present */
#define __DTCM_PRESENT            0U         /*!< Set to 1 if DTCM is present */

#include "core_cm4.h"                        /* Processor and core peripherals */
#include "system_ln882h.h"                   /* System Header */

#include <stddef.h>

#define CRITICAL_SECT_START()   do{                                            \
                                       uint32_t __irq_mask = __get_PRIMASK(); \
                                       __disable_irq();
#define CRITICAL_SECT_END()               if(__irq_mask==0) {__enable_irq();}    \
                                   }while(0);

#define GLOBAL_INT_START()      __enable_irq()
#define GLOBAL_INT_STOP()       __disable_irq()
                                    
#define GLOBAL_INT_DISABLE()    CRITICAL_SECT_START()
#define GLOBAL_INT_RESTORE()    CRITICAL_SECT_END()

/* -------  Start of section using anonymous unions and disabling warnings  ------- */
#if   defined (__CC_ARM)
  #pragma push
  #pragma anon_unions
#elif defined (__ICCARM__)
  #pragma language=extended
#elif defined(__ARMCC_VERSION) && (__ARMCC_VERSION >= 6010050)
  #pragma clang diagnostic push
  #pragma clang diagnostic ignored "-Wc11-extensions"
  #pragma clang diagnostic ignored "-Wreserved-id-macro"
#elif defined (__GNUC__)
  /* anonymous unions are enabled by default */
#elif defined (__TMS470__)
  /* anonymous unions are enabled by default */
#elif defined (__TASKING__)
  #pragma warning 586
#elif defined (__CSMC__)
  /* anonymous unions are enabled by default */
#else
  #warning Not supported compiler type
#endif


/* =========================================================================================================================== */
/* ================                            Device Specific Peripheral Section                             ================ */
/* =========================================================================================================================== */



/* --------  End of section using anonymous unions and disabling warnings  -------- */
#if   defined (__CC_ARM)
  #pragma pop
#elif defined (__ICCARM__)
  /* leave anonymous unions enabled */
#elif (defined(__ARMCC_VERSION) && (__ARMCC_VERSION >= 6010050))
  #pragma clang diagnostic pop
#elif defined (__GNUC__)
  /* anonymous unions are enabled by default */
#elif defined (__TMS470__)
  /* anonymous unions are enabled by default */
#elif defined (__TASKING__)
  #pragma warning restore
#elif defined (__CSMC__)
  /* anonymous unions are enabled by default */
#else
  #warning Not supported compiler type
#endif


/* =========================================================================================================================== */
/* ================                          Device Specific Peripheral Address Map                           ================ */
/* =========================================================================================================================== */
#if   defined (LN882H)
  #include "base_addr_ln882h.h"
  #include "mem_map_ln882h.h"
#else 
  #error Not supported device type
#endif

/**
 * @brief ln882h sdk version
 *
 * @Official Release URL: https://gitee.com/lightningsemi/ln882h
 *
 * LN882H_SDK_MAJOR_V:    0-255
 * LN882H_SDK_MINOR_V:    0-255
 */
#define LN882H_SDK_MAJOR_V        1
#define LN882H_SDK_MINOR_V        6

#define LN882H_SDK_RC_V           255/* 0: development, 255: release, 1-254: release candidates */

#define LN882H_SDK_RC_RELEASE     255
#define LN882H_SDK_RC_DEVELOPMENT 0

#define LN882H_SDK_VERSTR2(x)     #x
#define LN882H_SDK_VERSTR(x)      LN882H_SDK_VERSTR2(x)

#define LN882H_SDK_VERSION_IS_RELEASE     (LN882H_SDK_RC_V == LN882H_SDK_RC_RELEASE)
#define LN882H_SDK_VERSION_IS_DEVELOPMENT (LN882H_SDK_RC_V == LN882H_SDK_RC_DEVELOPMENT)
#define LN882H_SDK_VERSION_IS_RC          ((LN882H_SDK_RC_V != LN882H_SDK_RC_RELEASE) && (LN882H_SDK_RC_V != LN882H_SDK_RC_DEVELOPMENT))

#if LN882H_SDK_VERSION_IS_RELEASE
#define LN882H_SDK_VERSION_STRING_SUFFIX  ""
#elif LN882H_SDK_VERSION_IS_DEVELOPMENT
#define LN882H_SDK_VERSION_STRING_SUFFIX  "_alpha"
#else
#define LN882H_SDK_VERSION_STRING_SUFFIX  "_rc" LN882H_SDK_VERSTR(LN882H_SDK_RC_V)
#endif

#define LN882H_SDK_VERSION          ((uint32_t)(LN882H_SDK_MAJOR_V) << 24 |     \
                                     (uint32_t)(LN882H_SDK_MINOR_V) << 16  |    \
                                     (uint32_t)(LN882H_SDK_RC_V))
#define LN882H_SDK_VERSION_STRING   LN882H_SDK_VERSTR(LN882H_SDK_MAJOR_V) "."   \
                                    LN882H_SDK_VERSTR(LN882H_SDK_MINOR_V)       \
                                    LN882H_SDK_VERSION_STRING_SUFFIX
#define LN882H_SDK_BUILD_DATE_TIME  ""__DATE__" "__TIME__


#ifdef __cplusplus
}
#endif

#endif  /* __LN882H_H__ */
